In [2]:
import os
os.chdir("../")
%pwd

'c:\\Users\\abhis\\Desktop\\MLProjects\\Movie Recommender'

In [3]:
import pandas as pd

movie_df = pd.read_csv('artifacts/data_ingestion/ml-latest/movies.csv')
links_df = pd.read_csv('artifacts/data_ingestion/ml-latest/links.csv')
ratings_df = pd.read_csv('artifacts/data_ingestion/ml-latest/ratings.csv')
# tags_df = pd.read_csv('artifacts/data_ingestion/ml-latest/tags.csv')
# genome_scores_df = pd.read_csv('artifacts/data_ingestion/ml-latest/genome-scores.csv')
# genome_tags_df = pd.read_csv('artifacts/data_ingestion/ml-latest/genome-tags.csv')


In [4]:
# print('genome_tags_df \n',genome_tags_df.head(5))
# print('genome_scores_df \n',genome_scores_df.head(5))
# print('tags_df \n',tags_df.head())
print('ratings_df \n',ratings_df.head())
print('links_df \n',links_df.head())
print('movie_df \n',movie_df.head())

ratings_df 
    userId  movieId  rating   timestamp
0       1      307     3.5  1256677221
1       1      481     3.5  1256677456
2       1     1091     1.5  1256677471
3       1     1257     4.5  1256677460
4       1     1449     4.5  1256677264
links_df 
    movieId  imdbId   tmdbId
0        1  114709    862.0
1        2  113497   8844.0
2        3  113228  15602.0
3        4  114885  31357.0
4        5  113041  11862.0
movie_df 
    movieId                               title  \
0        1                    Toy Story (1995)   
1        2                      Jumanji (1995)   
2        3             Grumpier Old Men (1995)   
3        4            Waiting to Exhale (1995)   
4        5  Father of the Bride Part II (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                   Adventure|Children|Fantasy  
2                               Comedy|Romance  
3                         Comedy|Drama|Romance  
4                

In [5]:
# print('genome_tags_df \n',genome_tags_df.shape)
# print('genome_scores_df \n',genome_scores_df.shape)
# print('tags_df \n',tags_df.shape)
print('ratings_df \n',ratings_df.shape)
print('links_df \n',links_df.shape)
print('movie_df \n',movie_df.shape)

ratings_df 
 (27753444, 4)
links_df 
 (58098, 3)
movie_df 
 (58098, 3)


We don't want genome-scores and genome-tags and tags for this calculation. We will skip these three files.

Let's combine the other 3 files.

In [6]:
# first links_df and movie_df

links_movie_df = movie_df.merge(links_df, on='movieId')

In [7]:
links_movie_df.head(10)

movieId                               title  \
0        1                    Toy Story (1995)   
1        2                      Jumanji (1995)   
2        3             Grumpier Old Men (1995)   
3        4            Waiting to Exhale (1995)   
4        5  Father of the Bride Part II (1995)   
5        6                         Heat (1995)   
6        7                      Sabrina (1995)   
7        8                 Tom and Huck (1995)   
8        9                 Sudden Death (1995)   
9       10                    GoldenEye (1995)   

                                        genres  imdbId   tmdbId  
0  Adventure|Animation|Children|Comedy|Fantasy  114709    862.0  
1                   Adventure|Children|Fantasy  113497   8844.0  
2                               Comedy|Romance  113228  15602.0  
3                         Comedy|Drama|Romance  114885  31357.0  
4                                       Comedy  113041  11862.0  
5                        Action|Crime|Thriller  113277    949.0  
6                               Comedy|Romance  114319  11860.0  
7                           Adventure|Children  112302  45325.0  
8                                       Action  114576   9091.0  
9                    Action|Adventure|Thriller  113189    710.0

In [8]:
## Drop nan values

links_movie_df.dropna(inplace=True)

In [9]:
links_movie_df['tmdbId'] = links_movie_df['tmdbId'].astype(int)

In [10]:
links_movie_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 57917 entries, 0 to 58097
Data columns (total 5 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   movieId  57917 non-null  int64 
 1   title    57917 non-null  object
 2   genres   57917 non-null  object
 3   imdbId   57917 non-null  int64 
 4   tmdbId   57917 non-null  int32 
dtypes: int32(1), int64(2), object(2)
memory usage: 2.4+ MB


In [11]:
# remove timestamp from ratings and tag

ratings_df.drop(columns='timestamp', inplace=True)
ratings_df.head()

userId  movieId  rating
0       1      307     3.5
1       1      481     3.5
2       1     1091     1.5
3       1     1257     4.5
4       1     1449     4.5

In [12]:
ratings_df['userId'].nunique()

283228

In [13]:
ratings_df['movieId'].nunique()

53889

In [14]:
ratings_df['userId'].value_counts()

userId
123100    23715
117490     9279
134596     8381
212343     7884
242683     7515
          ...  
188125        1
117282        1
127062        1
241836        1
265726        1
Name: count, Length: 283228, dtype: int64

In [15]:
ratings_df['movieId'].value_counts()

movieId
318       97999
356       97040
296       92406
593       87899
2571      84545
          ...  
165077        1
172453        1
192975        1
167092        1
142855        1
Name: count, Length: 53889, dtype: int64

So there are 283228 unique users who rated the movies

In [16]:
# merge rating and movie_links

# movie_links_rating = pd.merge(left=links_movie_df, right=ratings_df, on='movieId')

In [17]:
# movie_links_rating.shape

In [18]:
# movie_links_rating.head()

In [24]:
#rearrange

# movie_links_rating = movie_links_rating[["userId","movieId", "title", "rating", "genres", "imdbId", "tmdbId"]]

In [25]:
# movie_links_rating.info()

In [26]:
# movie_links_rating.isnull().sum()

In [29]:
# API_KEY = os.environ.get('TMDB_API_KEY')


# def get_movie_details(movie_id):
#     movie_url = f'https://api.themoviedb.org/3/movie/{movie_id}?api_key={API_KEY}'

#     http = urllib3.PoolManager()
#     response = http.request('GET', movie_url)

#     if response.status == 200:
#         movie_data = json.loads(response.data.decode('utf-8'))

#         # Extracting the desired fields
#         genres = [genre['name'] for genre in movie_data['genres']]
#         overview = movie_data['overview']
#         popularity = movie_data['popularity']
#         poster_path = movie_data['poster_path']
#         vote_average = movie_data['vote_average']
#         vote_count = movie_data['vote_count']
#     else:
#         print('Error occurred while requesting data:', response.status)

#     # Extracting the director's name (assuming it is the first crew member with job "Director")
#     credits_url = f'https://api.themoviedb.org/3/movie/{movie_id}/credits?api_key={API_KEY}'
#     credits_response = http.request('GET', credits_url)

#     if credits_response.status == 200:
#         credits_data = json.loads(credits_response.data.decode('utf-8'))
#         crew = credits_data['crew']
#         director = next((member['name'] for member in crew if member['job'] == 'Director'), None)
#     else:
#         director = None

#     # Requesting the keywords endpoint
#     keywords_url = f'https://api.themoviedb.org/3/movie/{movie_id}/keywords?api_key={API_KEY}'
#     keywords_response = http.request('GET', keywords_url)

#     if keywords_response.status == 200:
#         keywords_data = json.loads(keywords_response.data.decode('utf-8'))
#         keywords = [keyword['name'] for keyword in keywords_data['keywords']]
#     else:
#         keywords = []

#     data = {
#         "genres": genres,
#         "overview": overview,
#         "popularity": popularity,
#         "poster_path": poster_path,
#         "vote_average": vote_average,
#         "vote_count": vote_count,
#         "director": director,
#         "keywords": keywords
#     }

#     return data

In [15]:
links_df.isnull().sum()

movieId      0
imdbId       0
tmdbId     181
dtype: int64

In [19]:
# let's fill the tmdbid first 

import json
import urllib.request

API_KEY = os.environ.get('TMDB_API_KEY')

def get_tmdb_id(imdb_id):
    full_imdb_id = f'tt0{int(imdb_id)}'
    find_url = f'https://api.themoviedb.org/3/find/{full_imdb_id}?api_key={API_KEY}&external_source=imdb_id'
    
    try:
        with urllib.request.urlopen(find_url) as response:
            data = json.load(response)
            movie_results = data.get('movie_results', [])
            if movie_results:
                tmdb_id = movie_results[0]['id']
                return tmdb_id
            else:
                print('No movie found for the given IMDb ID.',imdb_id)
    except urllib.error.HTTPError as e:
        print('Error occurred while requesting data:', e.code)

# Example usage
imdb_id = '114709'
tmdb_id = get_tmdb_id(imdb_id)
print('TMDB ID:', tmdb_id)

TMDB ID: 862


In [26]:
# import numpy as np

# # Assuming you have a pandas DataFrame named 'links_df' with 'imdbId' and 'tmdbId' columns

# # Function to replace missing TMDB IDs using get_tmdb_id function
# def replace_missing_tmdb_ids(row):
#     if pd.isnull(row['tmdbId']):
#         imdb_id = row['imdbId']
#         tmdb_id = get_tmdb_id(imdb_id)
#         return int(tmdb_id) if not pd.isnull(tmdb_id) else np.nan
#     else:
#         return int(row['tmdbId']) if not pd.isnull(row['tmdbId']) else np.nan

# # Apply the function to replace missing TMDB IDs
# links_df['tmdbId'] = links_df.apply(replace_missing_tmdb_ids, axis=1)

# # Print the updated DataFrame
# print(links_df.isnull().sum())

In [16]:
links_movie_df.shape

(57917, 5)

Data scraping from Tmdb to get some info like "genres": genres,
        "overview": overview,
        "popularity": popularity,
        "poster_path": poster_path,
        "vote_average": vote_average,
        "vote_count": vote_count,
        "director": director,
        "keywords": keywords

In [20]:
import os
import json
import urllib.request
import numpy as np
from tqdm import tqdm

API_KEY = os.environ.get('TMDB_API_KEY')
CACHE_DIR = 'artifacts/cached_data'

def fetch_movie_details_from_api(tmdb_id):
    movie_url = f'https://api.themoviedb.org/3/movie/{tmdb_id}?api_key={API_KEY}'
    credits_url = f'https://api.themoviedb.org/3/movie/{tmdb_id}/credits?api_key={API_KEY}'
    keywords_url = f'https://api.themoviedb.org/3/movie/{tmdb_id}/keywords?api_key={API_KEY}'

    movie_data = {}

    try:
        with urllib.request.urlopen(movie_url) as response:
            movie_data = json.loads(response.read().decode('utf-8'))

        with urllib.request.urlopen(credits_url) as response:
            credits_data = json.loads(response.read().decode('utf-8'))
            crew = credits_data.get('crew', [])
            director = next((member['name'] for member in crew if member.get('job') == 'Director'), np.nan)
            movie_data['director'] = director

        with urllib.request.urlopen(keywords_url) as response:
            keywords_data = json.loads(response.read().decode('utf-8'))
            keywords = [keyword['name'] for keyword in keywords_data.get('keywords', [])]
            movie_data['keywords'] = keywords

    except urllib.error.HTTPError as e:
        print(f'Error occurred while requesting movie data {tmdb_id}:', e.code)

    return movie_data

def fetch_batch_movie_details(tmdb_ids):
    movie_data_list = []
    for tmdb_id in tmdb_ids:
        cache_file = os.path.join(CACHE_DIR, f'{tmdb_id}.json')

        try:
            with open(cache_file, 'r') as file:
                movie_data = json.load(file)
        except FileNotFoundError:
            # If the cache file does not exist, fetch the movie details from the API
            movie_data = fetch_movie_details_from_api(tmdb_id)
            # Cache the movie details
            with open(cache_file, 'w') as file:
                json.dump(movie_data, file)
        

        movie_data_list.append(movie_data)
    

    return movie_data_list

def get_movie_details(tmdb_id):
    cache_file = os.path.join(CACHE_DIR, f'{tmdb_id}.json')

    try:
        with open(cache_file, 'r') as file:
            movie_data = json.load(file)
    except FileNotFoundError:
        # If the cache file does not exist, fetch the movie details from the API
        movie_data = fetch_movie_details_from_api(tmdb_id)
        # Cache the movie details
        with open(cache_file, 'w') as file:
            json.dump(movie_data, file)

    # Extract the desired fields from the movie data
    genres = [genre['name'] for genre in movie_data.get('genres', [])]
    overview = movie_data.get('overview', '')
    popularity = movie_data.get('popularity', 0)
    poster_path = movie_data.get('poster_path', '')
    vote_average = movie_data.get('vote_average', 0)
    vote_count = movie_data.get('vote_count', 0)
    director = movie_data.get('director', np.nan)
    keywords = movie_data.get('keywords', np.nan)

    data = {
        "genres": genres,
        "overview": overview,
        "popularity": popularity,
        "poster_path": poster_path,
        "vote_average": vote_average,
        "vote_count": vote_count,
        "director": director,
        "keywords": keywords
    }

    return data

def get_batch_movie_details(tmdb_ids):
    os.makedirs(CACHE_DIR, exist_ok=True)  # Create the cache directory if it doesn't exist

    movie_data_list = fetch_batch_movie_details(tmdb_ids)

    result = []
    for movie_data in movie_data_list:
        try:
            tmdb_id = movie_data['id']
        except KeyError:
            tmdb_id = None
        data = get_movie_details(tmdb_id)
        result.append(data)
    return result


In [24]:
tmdb_ids = links_movie_df.tmdbId.values

In [26]:

movie_data = get_batch_movie_details(tmdb_ids)

# get_batch_movie_details(tmdb_ids = tmdb_ids)

In [27]:
len(movie_data)

57917

In [21]:
movie_data[:2]

[{'genres': ['Animation', 'Adventure', 'Family', 'Comedy'],
  'overview': "Led by Woody, Andy's toys live happily in his room until Andy's birthday brings Buzz Lightyear onto the scene. Afraid of losing his place in Andy's heart, Woody plots against Buzz. But when circumstances separate Buzz and Woody from their owner, the duo eventually learns to put aside their differences.",
  'popularity': 101.402,
  'poster_path': '/uXDfjJbdP4ijW5hWSBrPrlKpxab.jpg',
  'vote_average': 8.0,
  'vote_count': 16771,
  'director': 'John Lasseter',
  'keywords': ['martial arts',
   'jealousy',
   'friendship',
   'bullying',
   'elementary school',
   'friends',
   'rivalry',
   'rescue',
   'computer animation',
   'mission',
   'anthropomorphism',
   'buddy',
   'walkie talkie',
   'toy car',
   'boy next door',
   'new toy',
   'neighborhood',
   'toy comes to life',
   'resourcefulness']},
 {'genres': ['Adventure', 'Fantasy', 'Family'],
  'overview': "When siblings Judy and Peter discover an enchante

In [34]:
tmdb_ids

array([   862,   8844,  15602, ..., 422666, 454439, 540871])

In [28]:
links_movie_df.head()

movieId                               title  \
0        1                    Toy Story (1995)   
1        2                      Jumanji (1995)   
2        3             Grumpier Old Men (1995)   
3        4            Waiting to Exhale (1995)   
4        5  Father of the Bride Part II (1995)   

                                        genres  imdbId  tmdbId  
0  Adventure|Animation|Children|Comedy|Fantasy  114709     862  
1                   Adventure|Children|Fantasy  113497    8844  
2                               Comedy|Romance  113228   15602  
3                         Comedy|Drama|Romance  114885   31357  
4                                       Comedy  113041   11862

In [29]:
get_movie_details(None)

{'genres': [],
 'overview': '',
 'popularity': 0,
 'poster_path': '',
 'vote_average': 0,
 'vote_count': 0,
 'director': nan,
 'keywords': nan}

In [36]:
# movie_links_rating.head()

In [46]:
tmdb_df = pd.DataFrame(movie_data)
tmdb_df['tmdbId'] = tmdb_ids

In [47]:
tmdb_df.shape, links_movie_df.shape

((57917, 9), (57917, 5))

In [48]:
tmdb_df.head()

genres  \
0  [Animation, Adventure, Family, Comedy]   
1            [Adventure, Fantasy, Family]   
2                       [Romance, Comedy]   
3                [Comedy, Drama, Romance]   
4                        [Comedy, Family]   

                                            overview  popularity  \
0  Led by Woody, Andy's toys live happily in his ...      94.409   
1  When siblings Judy and Peter discover an encha...      19.462   
2  A family wedding reignites the ancient feud be...      11.106   
3  Cheated on, mistreated and stepped on, the wom...      13.606   
4  Just when George Banks has recovered from his ...      14.332   

                        poster_path  vote_average  vote_count  \
0  /uXDfjJbdP4ijW5hWSBrPrlKpxab.jpg         7.968       16793   
1  /vgpXmVaVyUL7GGiDeiK1mKEKzcX.jpg         7.200        9651   
2  /1FSXpj5e8l4KH6nVFO5SPUeraOt.jpg         6.473         330   
3  /kJokIbVDkd6Ywp7IONv8xgfiES7.jpg         6.241         135   
4  /rj4LBtwQ0uGrpBnCELr716Qo3mw.jpg         6.246         644   

          director                                           keywords  tmdbId  
0    John Lasseter  [martial arts, jealousy, friendship, bullying,...     862  
1     Joe Johnston  [giant insect, board game, jungle, disappearan...    8844  
2    Howard Deutch  [fishing, halloween, sequel, old man, best fri...   15602  
3  Forest Whitaker  [based on novel or book, interracial relations...   31357  
4    Charles Shyer  [parent child relationship, baby, midlife cris...   11862

In [49]:
links_movie_df.head()

movieId                               title  \
0        1                    Toy Story (1995)   
1        2                      Jumanji (1995)   
2        3             Grumpier Old Men (1995)   
3        4            Waiting to Exhale (1995)   
4        5  Father of the Bride Part II (1995)   

                                        genres  imdbId  tmdbId  
0  Adventure|Animation|Children|Comedy|Fantasy  114709     862  
1                   Adventure|Children|Fantasy  113497    8844  
2                               Comedy|Romance  113228   15602  
3                         Comedy|Drama|Romance  114885   31357  
4                                       Comedy  113041   11862

In [50]:
links_movie_df_merged = pd.merge(links_movie_df, tmdb_df, on = 'tmdbId')

In [51]:
links_movie_df_merged[links_movie_df_merged['title'] == 'Dark Knight, The (2008)']

movieId                    title                 genres_x  imdbId  \
12581    58559  Dark Knight, The (2008)  Action|Crime|Drama|IMAX  468569   

       tmdbId                          genres_y  \
12581     155  [Drama, Action, Crime, Thriller]   

                                                overview  popularity  \
12581  Batman raises the stakes in his war on crime. ...       82.74   

                            poster_path  vote_average  vote_count  \
12581  /qJ2tW6WMUDux911r6m7haRef0WH.jpg         8.509       29859   

                director                                           keywords  
12581  Christopher Nolan  [crime fighter, secret identity, anti hero, sc...

In [33]:
links_movie_df_merged.shape

(57736, 13)

In [34]:
links_movie_df_merged.info()

<class 'pandas.core.frame.DataFrame'>
Index: 57736 entries, 0 to 57916
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   movieId       57736 non-null  int64  
 1   title         57736 non-null  object 
 2   genres_x      57736 non-null  object 
 3   imdbId        57736 non-null  int64  
 4   tmdbId        57736 non-null  int32  
 5   genres_y      57736 non-null  object 
 6   overview      57736 non-null  object 
 7   popularity    57736 non-null  float64
 8   poster_path   56364 non-null  object 
 9   vote_average  57736 non-null  float64
 10  vote_count    57736 non-null  int64  
 11  director      56643 non-null  object 
 12  keywords      56959 non-null  object 
dtypes: float64(2), int32(1), int64(3), object(7)
memory usage: 5.9+ MB


In [35]:
mask = links_movie_df_merged.isnull().any(axis=1)
rows_with_nan = links_movie_df_merged[mask]
len(rows_with_nan)

2357

In [36]:
57736-2357


55379

In [37]:
links_movie_df_merged.dropna(axis = 0, inplace=True)

In [38]:
links_movie_df_merged.isnull().sum()

movieId         0
title           0
genres_x        0
imdbId          0
tmdbId          0
genres_y        0
overview        0
popularity      0
poster_path     0
vote_average    0
vote_count      0
director        0
keywords        0
dtype: int64

In [39]:
links_movie_df_merged.info()

<class 'pandas.core.frame.DataFrame'>
Index: 55379 entries, 0 to 57916
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   movieId       55379 non-null  int64  
 1   title         55379 non-null  object 
 2   genres_x      55379 non-null  object 
 3   imdbId        55379 non-null  int64  
 4   tmdbId        55379 non-null  int32  
 5   genres_y      55379 non-null  object 
 6   overview      55379 non-null  object 
 7   popularity    55379 non-null  float64
 8   poster_path   55379 non-null  object 
 9   vote_average  55379 non-null  float64
 10  vote_count    55379 non-null  int64  
 11  director      55379 non-null  object 
 12  keywords      55379 non-null  object 
dtypes: float64(2), int32(1), int64(3), object(7)
memory usage: 5.7+ MB


genres from tmdb is more detailed so we drop the Movielens genre

In [40]:
links_movie_df_merged.drop(labels='genres_x', axis=1, inplace=True)

In [41]:
links_movie_df_merged.head()

movieId                               title  imdbId  tmdbId  \
0        1                    Toy Story (1995)  114709     862   
1        2                      Jumanji (1995)  113497    8844   
2        3             Grumpier Old Men (1995)  113228   15602   
3        4            Waiting to Exhale (1995)  114885   31357   
4        5  Father of the Bride Part II (1995)  113041   11862   

                                 genres_y  \
0  [Animation, Adventure, Family, Comedy]   
1            [Adventure, Fantasy, Family]   
2                       [Romance, Comedy]   
3                [Comedy, Drama, Romance]   
4                        [Comedy, Family]   

                                            overview  popularity  \
0  Led by Woody, Andy's toys live happily in his ...     101.402   
1  When siblings Judy and Peter discover an encha...      16.794   
2  A family wedding reignites the ancient feud be...       9.856   
3  Cheated on, mistreated and stepped on, the wom...      11.498   
4  Just when George Banks has recovered from his ...      14.211   

                        poster_path  vote_average  vote_count  \
0  /uXDfjJbdP4ijW5hWSBrPrlKpxab.jpg         8.000       16771   
1  /v2XHtmVqpERPy0HA1y9wltoeEgW.jpg         7.238        9636   
2  /1FSXpj5e8l4KH6nVFO5SPUeraOt.jpg         6.470         328   
3  /kJokIbVDkd6Ywp7IONv8xgfiES7.jpg         6.272         134   
4  /rj4LBtwQ0uGrpBnCELr716Qo3mw.jpg         6.250         642   

          director                                           keywords  
0    John Lasseter  [martial arts, jealousy, friendship, bullying,...  
1     Joe Johnston  [giant insect, board game, jungle, disappearan...  
2    Howard Deutch  [fishing, halloween, sequel, old man, best fri...  
3  Forest Whitaker  [based on novel or book, interracial relations...  
4    Charles Shyer  [parent child relationship, baby, midlife cris...

In [42]:
links_movie_df_merged.rename(columns = {'genres_y': 'genres'}, inplace=True)

In [43]:
links_movie_df_merged.head()

movieId                               title  imdbId  tmdbId  \
0        1                    Toy Story (1995)  114709     862   
1        2                      Jumanji (1995)  113497    8844   
2        3             Grumpier Old Men (1995)  113228   15602   
3        4            Waiting to Exhale (1995)  114885   31357   
4        5  Father of the Bride Part II (1995)  113041   11862   

                                   genres  \
0  [Animation, Adventure, Family, Comedy]   
1            [Adventure, Fantasy, Family]   
2                       [Romance, Comedy]   
3                [Comedy, Drama, Romance]   
4                        [Comedy, Family]   

                                            overview  popularity  \
0  Led by Woody, Andy's toys live happily in his ...     101.402   
1  When siblings Judy and Peter discover an encha...      16.794   
2  A family wedding reignites the ancient feud be...       9.856   
3  Cheated on, mistreated and stepped on, the wom...      11.498   
4  Just when George Banks has recovered from his ...      14.211   

                        poster_path  vote_average  vote_count  \
0  /uXDfjJbdP4ijW5hWSBrPrlKpxab.jpg         8.000       16771   
1  /v2XHtmVqpERPy0HA1y9wltoeEgW.jpg         7.238        9636   
2  /1FSXpj5e8l4KH6nVFO5SPUeraOt.jpg         6.470         328   
3  /kJokIbVDkd6Ywp7IONv8xgfiES7.jpg         6.272         134   
4  /rj4LBtwQ0uGrpBnCELr716Qo3mw.jpg         6.250         642   

          director                                           keywords  
0    John Lasseter  [martial arts, jealousy, friendship, bullying,...  
1     Joe Johnston  [giant insect, board game, jungle, disappearan...  
2    Howard Deutch  [fishing, halloween, sequel, old man, best fri...  
3  Forest Whitaker  [based on novel or book, interracial relations...  
4    Charles Shyer  [parent child relationship, baby, midlife cris...

In [44]:
links_movie_df_merged.columns

Index(['movieId', 'title', 'imdbId', 'tmdbId', 'genres', 'overview',
       'popularity', 'poster_path', 'vote_average', 'vote_count', 'director',
       'keywords'],
      dtype='object')

In [47]:
genre_column

0        Adventure|Animation|Children|Comedy|Fantasy
1                         Adventure|Children|Fantasy
2                                     Comedy|Romance
3                               Comedy|Drama|Romance
4                                             Comedy
                            ...                     
58093                             (no genres listed)
58094                                         Comedy
58095                                          Drama
58096                  Adventure|Drama|Horror|Sci-Fi
58097                             Action|Crime|Drama
Name: genres, Length: 57917, dtype: object

In [48]:
# Extract the 'genres' column
genre_column = links_movie_df_merged['genres']

# Create an empty set for unique genres
unique_genres = set()

# Iterate over each row in the genre column
for genres_list in genre_column:
    
    # Add each genre to the set
    unique_genres.update(genres_list)

# Get the total number of unique genres
total_genres = len(unique_genres)

print(total_genres)  # Output: 7

19


In [50]:
list(unique_genres)

['Action',
 'Mystery',
 'Western',
 'Romance',
 'Music',
 'War',
 'Fantasy',
 'Horror',
 'History',
 'Documentary',
 'Adventure',
 'TV Movie',
 'Science Fiction',
 'Animation',
 'Comedy',
 'Drama',
 'Family',
 'Thriller',
 'Crime']

In [51]:
final_df = links_movie_df_merged.merge(ratings_df, on='movieId')

In [52]:
final_df.shape,links_movie_df_merged.shape,ratings_df.shape

((27372632, 14), (55379, 12), (27753444, 3))

In [53]:
final_df.head()

movieId             title  imdbId  tmdbId  \
0        1  Toy Story (1995)  114709     862   
1        1  Toy Story (1995)  114709     862   
2        1  Toy Story (1995)  114709     862   
3        1  Toy Story (1995)  114709     862   
4        1  Toy Story (1995)  114709     862   

                                   genres  \
0  [Animation, Adventure, Family, Comedy]   
1  [Animation, Adventure, Family, Comedy]   
2  [Animation, Adventure, Family, Comedy]   
3  [Animation, Adventure, Family, Comedy]   
4  [Animation, Adventure, Family, Comedy]   

                                            overview  popularity  \
0  Led by Woody, Andy's toys live happily in his ...     101.402   
1  Led by Woody, Andy's toys live happily in his ...     101.402   
2  Led by Woody, Andy's toys live happily in his ...     101.402   
3  Led by Woody, Andy's toys live happily in his ...     101.402   
4  Led by Woody, Andy's toys live happily in his ...     101.402   

                        poster_path  vote_average  vote_count       director  \
0  /uXDfjJbdP4ijW5hWSBrPrlKpxab.jpg           8.0       16771  John Lasseter   
1  /uXDfjJbdP4ijW5hWSBrPrlKpxab.jpg           8.0       16771  John Lasseter   
2  /uXDfjJbdP4ijW5hWSBrPrlKpxab.jpg           8.0       16771  John Lasseter   
3  /uXDfjJbdP4ijW5hWSBrPrlKpxab.jpg           8.0       16771  John Lasseter   
4  /uXDfjJbdP4ijW5hWSBrPrlKpxab.jpg           8.0       16771  John Lasseter   

                                            keywords  userId  rating  
0  [martial arts, jealousy, friendship, bullying,...       4     4.0  
1  [martial arts, jealousy, friendship, bullying,...      10     5.0  
2  [martial arts, jealousy, friendship, bullying,...      14     4.5  
3  [martial arts, jealousy, friendship, bullying,...      15     4.0  
4  [martial arts, jealousy, friendship, bullying,...      22     4.0

In [55]:
final_df.columns

Index(['movieId', 'title', 'imdbId', 'tmdbId', 'genres', 'overview',
       'popularity', 'poster_path', 'vote_average', 'vote_count', 'director',
       'keywords', 'userId', 'rating'],
      dtype='object')

In [54]:
final_df.isnull().sum()

movieId         0
title           0
imdbId          0
tmdbId          0
genres          0
overview        0
popularity      0
poster_path     0
vote_average    0
vote_count      0
director        0
keywords        0
userId          0
rating          0
dtype: int64

In [56]:
final_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27372632 entries, 0 to 27372631
Data columns (total 14 columns):
 #   Column        Dtype  
---  ------        -----  
 0   movieId       int64  
 1   title         object 
 2   imdbId        int64  
 3   tmdbId        int32  
 4   genres        object 
 5   overview      object 
 6   popularity    float64
 7   poster_path   object 
 8   vote_average  float64
 9   vote_count    int64  
 10  director      object 
 11  keywords      object 
 12  userId        int64  
 13  rating        float64
dtypes: float64(3), int32(1), int64(4), object(6)
memory usage: 2.8+ GB
